In [ ]:
!pip install chromadb datasets

---------------------------------
#### ChromaDB

- Using Hugging Face datasets library
- understaning chromaDB's storage options (Flat, Inverted etc...)
- understanding distance function
- notice the time taken to create the embeddings
-------------------------------

In [1]:
from datasets import load_dataset

#### Dataset Overview:
- **Name**: MedMCQA
- **Domain**: Medical Question Answering
- **Type**: Multiple-Choice QA (MCQA)
- **Size**: Large-scale dataset with tens of thousands of questions.
- **Purpose**: The dataset is used to train and evaluate models on medical question answering, helping to develop AI systems that can assist with medical exam preparation and knowledge extraction.
  
#### Structure:
- The dataset consists of **multiple-choice questions** with four possible answers.
- For each question, there is a correct answer and additional metadata.

#### Key Features:
1. **Question** (`question`): The medical question in text form, usually related to clinical scenarios, medical knowledge, or diagnostic reasoning.
2. **Answer Choices** (`opa`, `opb`, `opc`, `opd`): Four possible options (A, B, C, and D) from which the model is expected to pick the correct answer.
3. **Correct Answer** (`cop`): The correct option among the four provided choices.
4. **Explanations** (`explanation`): Often, there is an explanation provided to justify the correct answer, offering insight into the reasoning process behind the choice.
5. **Metadata**: Additional fields such as the question type, difficulty, subject area, and any specific medical domain or context.

#### Example Entry:
- **Question**: "Which of the following is the most common cause of fever in children under the age of five?"
- **Option A**: Viral infection
- **Option B**: Bacterial infection
- **Option C**: Fungal infection
- **Option D**: Parasitic infection
- **Correct Answer**: A
- **Explanation**: Viral infections are the most common cause of fever in children under five, followed by bacterial infections.

#### Use Cases:
- **AI Research in Healthcare**: MedMCQA is valuable for developing models that answer complex, domain-specific questions related to medicine.
- **Exam Preparation Tools**: It can be used to create intelligent tutoring systems that assist medical students in preparing for exams.
- **Medical QA Systems**: The dataset is essential for building AI systems capable of understanding medical queries and providing accurate answers, a crucial application in clinical decision support.

#### Tasks:
- **Classification**: The model is tasked with classifying the correct answer among four possible options.
- **Natural Language Processing (NLP)**: Understanding and reasoning over medical text is key to performing well on this dataset.


In [2]:
dataset = load_dataset("medmcqa", split="train")

In [3]:
type(dataset)

datasets.arrow_dataset.Dataset

In [4]:
dataset.shape

(182822, 11)

In [5]:
from pprint import pprint

In [6]:
# View the first 5 examples
print("First 5 examples:\n")
pprint(dataset[:5])  # Using pprint for better readability

First 5 examples:

{'choice_type': ['single', 'single', 'multi', 'multi', 'single'],
 'cop': [2, 2, 3, 0, 1],
 'exp': ['Chronic urethral obstruction because of urinary calculi, prostatic '
         'hyperophy, tumors, normal pregnancy, tumors, uterine prolapse or '
         'functional disorders cause hydronephrosis which by definition is '
         'used to describe dilatation of renal pelvis and calculus associated '
         'with progressive atrophy of the kidney due to obstruction to the '
         'outflow of urine Refer Robbins 7yh/9,1012,9/e. P950',
         "Ans. (c) Vitamin B12 Ref: Harrison's 19th ed. P 640* Vitamin B12 "
         '(Cobalamin) is synthesized solely by microorganisms.* In humans, the '
         'only source for humans is food of animal origin, e.g., meat, fish, '
         'and dairy products.* Vegetables, fruits, and other foods of '
         "nonanimal origin doesn't contain Vitamin B12 .* Daily requirements "
         'of vitamin Bp is about 1-3 pg. Body st

In [7]:
# Get information about the dataset's features (columns)
print("\nDataset features:\n")
pprint(dataset.features)  # Using pprint for better readability


Dataset features:

{'choice_type': Value('string'),
 'cop': ClassLabel(names=['a', 'b', 'c', 'd']),
 'exp': Value('string'),
 'id': Value('string'),
 'opa': Value('string'),
 'opb': Value('string'),
 'opc': Value('string'),
 'opd': Value('string'),
 'question': Value('string'),
 'subject_name': Value('string'),
 'topic_name': Value('string')}


In [8]:
# Check the total number of examples
print("\nTotal number of examples:\n")
print(len(dataset))


Total number of examples:

182822


In [9]:
# View the first example with field names
print("\nFirst example with fields:\n")
pprint(dataset[:1])  # Using pprint for better readability


First example with fields:

{'choice_type': ['single'],
 'cop': [2],
 'exp': ['Chronic urethral obstruction because of urinary calculi, prostatic '
         'hyperophy, tumors, normal pregnancy, tumors, uterine prolapse or '
         'functional disorders cause hydronephrosis which by definition is '
         'used to describe dilatation of renal pelvis and calculus associated '
         'with progressive atrophy of the kidney due to obstruction to the '
         'outflow of urine Refer Robbins 7yh/9,1012,9/e. P950'],
 'id': ['e9ad821a-c438-4965-9f77-760819dfa155'],
 'opa': ['Hyperplasia'],
 'opb': ['Hyperophy'],
 'opc': ['Atrophy'],
 'opd': ['Dyplasia'],
 'question': ['Chronic urethral obstruction due to benign prismatic '
              'hyperplasia can lead to the following change in kidney '
              'parenchyma'],
 'subject_name': ['Anatomy'],
 'topic_name': ['Urinary tract']}


In [10]:
import pandas as pd
# Set Pandas to display entire column width
pd.set_option('display.max_colwidth', None)

In [11]:
# Convert to a Pandas DataFrame for easier inspection
df = dataset.to_pandas()
df.head()

,id,question,opa,opb,opc,opd,cop,choice_type,exp,subject_name,topic_name
0,e9ad821a-c438-4965-9f77-760819dfa155,Chronic urethral obstruction due to benign prismatic hyperplasia can lead to the following change in kidney parenchyma,Hyperplasia,Hyperophy,Atrophy,Dyplasia,2,single,"Chronic urethral obstruction because of urinary calculi, prostatic hyperophy, tumors, normal pregnancy, tumors, uterine prolapse or functional disorders cause hydronephrosis which by definition is used to describe dilatation of renal pelvis and calculus associated with progressive atrophy of the kidney due to obstruction to the outflow of urine Refer Robbins 7yh/9,1012,9/e. P950",Anatomy,Urinary tract
1,e3d3c4e1-4fb2-45e7-9f88-247cc8f373b3,Which vitamin is supplied from only animal source:,Vitamin C,Vitamin B7,Vitamin B12,Vitamin D,2,single,"Ans. (c) Vitamin B12 Ref: Harrison's 19th ed. P 640* Vitamin B12 (Cobalamin) is synthesized solely by microorganisms.* In humans, the only source for humans is food of animal origin, e.g., meat, fish, and dairy products.* Vegetables, fruits, and other foods of nonanimal origin doesn't contain Vitamin B12 .* Daily requirements of vitamin Bp is about 1-3 pg. Body stores are of the order of 2-3 mg, sufficient for 3-4 years if supplies are completely cut off.",Biochemistry,Vitamins and Minerals
2,5c38bea6-787a-44a9-b2df-88f4218ab914,All of the following are surgical options for morbid obesity except -,Adjustable gastric banding,Biliopancreatic diversion,Duodenal Switch,Roux en Y Duodenal By pass,3,multi,"Ans. is 'd' i.e., Roux en Y Duodenal Bypass Bariatric surgical procedures include:a. Vertical banded gastroplastyb. Adjustable gastric bandingc. Roux-en Y gastric bypass (Not - Roux-en Y Duodenal Bypass)d. Biliopancreatic diversione. Duodenal switcho The surgical treatment of morbid obesity is known as bariatric surgery.o Morbid obesity is defined as body mass index of 35 kg/m2 or more with obesity related comorbidity, or BMI of 40 kg/m2 or greater without comorbidity.o Bariatric operations produce weight loss as a result of 2 factors. One is restriction of oralintake. The other is malabsorbtion of ingested food.o Gastric restrictive procedures include Vertical banded gastroplasty & Adjustable gastric bandingo Malabsorbtive procedures include Biliopancreatic diversion, and Duodenal switcho Roux-en Y gastric bypass has features of both restriction and malabsorptionBariatric Operations: Mechanism of ActionRestrictiveVertical banded gastroplastyLaparoscopic adjustable gastric bandingLargely Restrictive/Mildly MalabsorptiveRoux-en-Y gastric bypassLargely Malabsorptive/Mildly RestrictiveBiliopancreatic diversionDuodenal switch",Surgery,Surgical Treatment Obesity
3,cdeedb04-fbe9-432c-937c-d53ac24475de,"Following endaerectomy on the right common carotid, a patient is found to be blind in the right eye. It is appears that a small thrombus embolized during surgery and lodged in the aery supplying the optic nerve. Which aery would be blocked?",Central aery of the retina,Infraorbital aery,Lacrimal aery,Nasociliary aretry,0,multi,"The central aery of the retina is a branch of the ophthalmic aery. It is the sole blood supply to the retina; it has no significant collateral circulation and blockage of this vessel leads to blindness. The branches of this aery are what you view during a fundoscopic exam. Note: The infraorbital aery is a branch of the maxillary aery. It comes through the infraorbital foramen, inferior to the eye. It supplies the maxillary sinus, the maxillary incisors, canine and premolar teeth, and the skin of the cheek below the orbit. The supraorbital aery is another branch of the maxillary aery. It comes through the supraorbital foramen and supplies blood to the muscles, skin and fascia of the forehead. The lacrimal aery is a branch of the ophthalmic aery that supplies the lacrimal gland. The nasociliary aery doesn't exist, but there is a nasociliary nerve that travels with the ophthalmic aery. Ref: Moon D.A., Forem

#### Data sanity checks

In [12]:
df.isna().sum()

id                  0
question            0
opa                 0
opb                 0
opc                 0
opd                 0
cop                 0
choice_type         0
exp             21953
subject_name        0
topic_name      95613
dtype: int64

In [13]:
# Count of missing values for each column
missing_count = df.isna().sum()

# Percentage of missing values for each column
missing_percentage = (missing_count / len(df)) * 100

# Create a summary DataFrame to show both counts and percentages
missing_summary = pd.DataFrame({
    'Missing Count': missing_count,
    'Missing Percentage (%)': missing_percentage
})

# Filter to show only columns with missing values
missing_summary = missing_summary[missing_summary['Missing Count'] > 0]
missing_summary

,Missing Count,Missing Percentage (%)
exp,21953,12.007855
topic_name,95613,52.298410


Columns `exp` and `topic_name` contain missing values, we will drop rows with missing values in both columns

In [14]:
# Dropping rows with missing values in 'exp' and 'topic_name' columns
df_cleaned = df.dropna(subset=['exp', 'topic_name'])
df_cleaned.shape

(87077, 11)

In [15]:
df_cleaned.isnull().sum()

id              0
question        0
opa             0
opb             0
opc             0
opd             0
cop             0
choice_type     0
exp             0
subject_name    0
topic_name      0
dtype: int64

Now your dataset has no missing values, we will convert dataframe back into a dataset.

In [16]:
#converting the pre-processed dataframe back to dataset
dataset_cleaned = dataset.from_pandas(df_cleaned)

#### Creating a Collection in Chroma

- **Purpose**: Chroma uses Collections to store embeddings, documents, and metadata.
- **Method**: Use the `createCollection()` method of the Chroma client.
- **Parameters**:
  - **Collection Name**: Provide a name for the collection.
  - **Optional Embedding Function**: 
    - You can specify an embedding function to generate embeddings from text.
    - If no embedding function is provided, the default embedding function is used.
- **Adding Data to the Collection**:
  - After creating a collection, you need to add your data to it.
  - For each data point, you can add the following:
    - **ID**: Unique identifier for the data point.
    - **Data**: The actual data (e.g., text or document).
    - **Metadata**: Any additional information related to the data point.


In [17]:
import chromadb

In [18]:
client = chromadb.Client()

In [19]:
collection = client.create_collection("MedicalQA")

In [20]:
# List all collections
collections = client.list_collections()
print([collection.name for collection in collections])

['MedicalQA']


In [21]:
# Now we will store the expert explanation field of first 10 questions from dataset into collection. 
collection.add(
    ids      = [str(i) for i in range(0, 10)],           # every entry in collection requires an ID, ID is a string
    documents= dataset["exp"][:10],                      # explanation field text, for first 10 questions only
    metadatas= [{"type": "exp"} for _ in range(0, 10)],) # metadata, additional information about the documents

In [22]:
collection.peek()

{'ids': ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'],
 'embeddings': array([[ 0.01779646,  0.02271976,  0.07299382, ...,  0.04815544,
          0.09588588,  0.02359383],
        [-0.03943812,  0.00844764, -0.11407645, ...,  0.0629646 ,
          0.01048637, -0.03838765],
        [ 0.0184587 ,  0.01128408, -0.05528871, ..., -0.0362584 ,
          0.00104199, -0.06918139],
        ...,
        [-0.01332261, -0.01707296, -0.04146404, ...,  0.02947679,
          0.0990802 ,  0.01221246],
        [ 0.03374663, -0.04208902, -0.01137699, ...,  0.05283229,
          0.09082187,  0.03532808],
        [ 0.02446118, -0.00422286,  0.03287291, ...,  0.07694613,
          0.00944527, -0.00467342]], shape=(10, 384)),
 'documents': ['Chronic urethral obstruction because of urinary calculi, prostatic hyperophy, tumors, normal pregnancy, tumors, uterine prolapse or functional disorders cause hydronephrosis which by definition is used to describe dilatation of renal pelvis and calculus associated w

#### How are the vectors created (automatically)

`Default Embedding Model in ChromaDB`

- **Model Name**: all-MiniLM-L6-v2
- **Type**: Sentence Transformers model
- **Functionality**: 
  - The all-MiniLM-L6-v2 model is used by default to create embeddings for sentences and documents.
  - This model can be utilized for a variety of tasks including semantic search, clustering, and classification.
- **Execution**: 
  - The embedding function runs locally on your machine.
  - Model files may need to be downloaded, which will happen automatically when you first run the embedding function.
 
e.g. `C:\Users\bhupe\.cache\torch\sentence_transformers\sentence-transformers_all-MiniLM-L6-v2`

- Embeddings: Generated for documents (text content).
- Metadata: Used to provide context or additional information but not embedded or transformed into vectors.

In [23]:
collection.get_model()

Collection(id=UUID('b2e24005-e602-4d60-b76e-960f369e5b5f'), name='MedicalQA', configuration_json={'hnsw': {'space': 'l2', 'ef_construction': 100, 'ef_search': 100, 'max_neighbors': 16, 'resize_factor': 1.2, 'sync_threshold': 1000}, 'spann': None, 'embedding_function': {'type': 'known', 'name': 'default', 'config': {}}}, serialized_schema={'defaults': {'string': {'fts_index': {'enabled': False, 'config': {}}, 'string_inverted_index': {'enabled': True, 'config': {}}}, 'float_list': {'vector_index': {'enabled': False, 'config': {'space': 'l2', 'embedding_function': {'type': 'known', 'name': 'default', 'config': {}}, 'source_key': '#document', 'hnsw': {'ef_construction': 100, 'max_neighbors': 16, 'ef_search': 100, 'num_threads': 16, 'batch_size': 100, 'sync_threshold': 1000, 'resize_factor': 1.2}}}}, 'sparse_vector': {'sparse_vector_index': {'enabled': False, 'config': {'embedding_function': {'type': 'unknown'}, 'bm25': False}}}, 'int': {'int_inverted_index': {'enabled': True, 'config': {}

#### all-MiniLM-L6-v2 Model Overview

- **Model Type**: Sentence Transformer
- **Architecture**: 
  - This model is based on the **MiniLM** architecture, which is a compact version of the BERT model, specifically designed for efficiency and effectiveness in generating sentence embeddings.

- **Dimensions**: 
  - The model produces **384-dimensional** embeddings, making it suitable for tasks where a balance between performance and resource efficiency is required.

- **Training Data**: 
  - all-MiniLM-L6-v2 is trained on a diverse range of datasets, primarily focused on sentence-level tasks, which enables it to capture semantic meaning effectively.

- **Use Cases**: 
  - **Semantic Search**: Finding documents or text snippets that are semantically similar to a query.
  - **Clustering**: Grouping similar sentences or documents based on their embeddings.
  - **Classification**: Classifying sentences or documents into predefined categories.
  - **Text Similarity**: Measuring the similarity between different pieces of text.

- **Performance**:
  - The model is optimized for **speed and memory efficiency**, making it suitable for real-time applications and deployment on devices with limited computational resources.

- **Embedding Generation**:
  - It generates embeddings that represent the semantic content of input sentences. This means that sentences with similar meanings will have embeddings that are close together in the embedding space.

### Key Points about BERT Models (For New AI Learners)

- **Understands Context:**
  - BERT reads the whole sentence at once, so it understands the meaning of each word based on the words around it.

- **Great for Many Tasks:**
  - BERT can be used for answering questions, finding similar sentences, classifying text, and more.

- **Pretrained on Lots of Text:**
  - BERT has already read a huge amount of text (like books and Wikipedia) before you use it, so it knows a lot about language.

- **Fine-tuning:**
  - You can teach BERT to do a specific job (like detecting spam) by giving it a small set of examples for that job.

- **Handles Both Directions:**
  - BERT looks at words before and after a target word, so it gets a better understanding than models that only read left-to-right or right-to-left.

- **Open Source and Widely Used:**
  - BERT is free to use and has become a standard tool in natural language processing (NLP).

- **Embeddings:**
  - BERT turns sentences or words into number lists (embeddings) that capture their meaning, which can be used for search, clustering, and more.

- **Improves Search and Chatbots:**
  - BERT helps computers understand questions and find the best answers, making search engines and chatbots smarter.

- **Easy to Use:**
  - Many libraries make it simple to use BERT with just a few lines of code.

In [24]:
# Peek at a specific row in the collection by ID
row_id   = str(1)  # Replace 0 with the desired row index (0 to 9 for the first 10 rows)
document = collection.get(row_id)  # Retrieve the document by its ID

document

{'ids': ['1'],
 'embeddings': None,
 'documents': ["Ans. (c) Vitamin B12 Ref: Harrison's 19th ed. P 640* Vitamin B12 (Cobalamin) is synthesized solely by microorganisms.* In humans, the only source for humans is food of animal origin, e.g., meat, fish, and dairy products.* Vegetables, fruits, and other foods of nonanimal origin doesn't contain Vitamin B12 .* Daily requirements of vitamin Bp is about 1-3 pg. Body stores are of the order of 2-3 mg, sufficient for 3-4 years if supplies are completely cut off."],
 'uris': None,
 'included': ['metadatas', 'documents'],
 'data': None,
 'metadatas': [{'type': 'exp'}]}

In [25]:
collection.peek(1)['embeddings'].shape

(1, 384)

In [26]:
query_text = '''Chronic urethral obstruction from various causes leads to hydronephrosis, 
characterized by renal pelvis dilation and kidney atrophy due to urine blockage.'''

In [27]:
results = collection.query(query_texts=query_text, n_results=3)

In [28]:
results

{'ids': [['0', '6', '7']],
 'embeddings': None,
 'documents': [['Chronic urethral obstruction because of urinary calculi, prostatic hyperophy, tumors, normal pregnancy, tumors, uterine prolapse or functional disorders cause hydronephrosis which by definition is used to describe dilatation of renal pelvis and calculus associated with progressive atrophy of the kidney due to obstruction to the outflow of urine Refer Robbins 7yh/9,1012,9/e. P950',
   'Abnormal vascular pattern include punctation, mosaicism and atypical vessels.',
   "PILONIDAL SINUS/DISEASE (Jeep Bottom; Driver's Bottom) Pilus--hair; Nidus--nest It is epithelium lined tract, situated sho distance behind the anus, containing hairs and unhealthy diseased granula- tion tissue. It is due to penetration of hairs through the skin into subcutaneous tissue. It forms granuloma/unhealthy granulation tissue in the deeper plane. It is of infective origin and occurs in sacral region between the buttocks, umbilicus, axilla. Other indic

In [29]:
query_texts = [
    'Urinary obstructions, like tumors or pregnancy, lead to hydronephrosis, which causes kidney dilation and atrophy due to urine blockage.',
    'Hydronephrosis results from chronic urinary obstructions such as tumors, pregnancy, or prostatic hyperplasia, causing kidney dilation and atrophy.',
    'Urethral obstruction from stones or tumors can cause hydronephrosis, leading to renal pelvis dilation and kidney atrophy due to blockage.',
    'Conditions like prostatic enlargement or tumors obstruct urine flow, causing hydronephrosis, characterized by renal pelvis dilation and kidney atrophy.',
    'Hydronephrosis occurs due to chronic urethral obstruction from calculi, tumors, or pregnancy, leading to kidney dilation and atrophy.'
]

In [30]:
for each_query in query_texts:
    results = collection.query(query_texts=each_query, n_results=1)

    print(each_query)
    print(results['documents'])
    print(results['distances'])

    print()

Urinary obstructions, like tumors or pregnancy, lead to hydronephrosis, which causes kidney dilation and atrophy due to urine blockage.
[['Chronic urethral obstruction because of urinary calculi, prostatic hyperophy, tumors, normal pregnancy, tumors, uterine prolapse or functional disorders cause hydronephrosis which by definition is used to describe dilatation of renal pelvis and calculus associated with progressive atrophy of the kidney due to obstruction to the outflow of urine Refer Robbins 7yh/9,1012,9/e. P950']]
[[0.413501501083374]]

Hydronephrosis results from chronic urinary obstructions such as tumors, pregnancy, or prostatic hyperplasia, causing kidney dilation and atrophy.
[['Chronic urethral obstruction because of urinary calculi, prostatic hyperophy, tumors, normal pregnancy, tumors, uterine prolapse or functional disorders cause hydronephrosis which by definition is used to describe dilatation of renal pelvis and calculus associated with progressive atrophy of the kidney

#### Distances

**1. Squared L2 Norm**, also known as **L2 Squared** or **Euclidean Distance Squared**, measures the distance between two points in space without the computational cost of a square root.

##### Formula:
For two vectors \( x \) and \( y \), the squared L2 norm is calculated as:
```python
# L2 Squared Formula:
import numpy as np

x = np.array([x1, x2, x3])  # example vector x
y = np.array([y1, y2, y3])  # example vector y

squared_l2 = np.sum((x - y)**2)
print(squared_l2)
```

**Key Points:**
- Avoids square root: Simplifies calculations by omitting the square root operation.
- Distance measure: Represents the squared distance between two points in Euclidean space.
- Common usage: Widely used in machine learning optimization problems, such as linear regression, where exact distances are not required.

**2. inner product**, also known as the **dot product** in Euclidean space, is a measure of similarity between two vectors. It returns a scalar value and is used in many machine learning and linear algebra applications.

##### Formula:
For two vectors \( x \) and \( y \) with \( n \) components, the inner product is calculated as:
$$
x \cdot y = \sum_{i=1}^n x_i y_i
$$

##### Python Example:
```python
import numpy as np

# Define two vectors x and y
x = np.array([x1, x2, x3])
y = np.array([y1, y2, y3])

# Calculate the inner product (dot product)
inner_product = np.dot(x, y)
print(inner_product)

```

**Key Points:**
- Similarity measure: If the inner product is large, the vectors are more aligned; if zero, the vectors are orthogonal.
- Applications: Used in machine learning, neural networks, and geometric computations.


**3. Cosine Similarity** is a metric used to measure the cosine of the angle between two non-zero vectors in an inner product space. It calculates the similarity between two vectors based on their direction, ignoring their magnitude.

For two vectors \( x \) and \( y \), cosine similarity is defined as:
$$
\text{cosine\_similarity}(x, y) = \frac{x \cdot y}{\|x\| \|y\|}
$$

Where:
- $( x \cdot y )$ is the inner product (dot product) of vectors \( x \) and \( y \).
- $( \|x\| ) and ( |y\| )$ are the Euclidean norms (L2 norms) of vectors \( x \) and \( y \).

```python
import numpy as np

# Define two vectors x and y
x = np.array([x1, x2, x3])
y = np.array([y1, y2, y3])

# Calculate cosine similarity
cosine_similarity = np.dot(x, y) / (np.linalg.norm(x) * np.linalg.norm(y))
print(cosine_similarity)

```

**Key Points:**

- Range: The result of cosine similarity lies between -1 and 1.

    - 1 indicates vectors are identical in direction.
    - 0 indicates orthogonality (no similarity).
    - -1 indicates vectors are diametrically opposed.
      
Applications: Commonly used in text analysis, recommendation systems, and other tasks involving high-dimensional data.


**4. Cosine Distance (1 - Cosine Similarity)**

The expression **1 - Cosine Similarity** is used to compute **Cosine Distance**, which measures the dissimilarity between two vectors. While cosine similarity assesses the similarity, cosine distance focuses on how far apart the vectors are.

For two vectors \( x \) and \( y \), the **Cosine Distance** is calculated as:
$$
\text{cosine\_distance}(x, y) = 1 - \frac{x \cdot y}{\|x\| \|y\|}
$$

Where:
- \( x \cdot y \) is the inner product (dot product) of vectors \( x \) and \( y \).
- \( \|x\| \) and \( \|y\| \) are the L2 norms of vectors \( x \) and \( y \).

- **Range**: The cosine distance lies between 0 and 2:
  - **0**: Vectors are identical.
  - **1**: Vectors are orthogonal (no similarity).
  - **2**: Vectors are diametrically opposed.
- **Applications**: Commonly used in clustering, similarity measures, and distance-based algorithms.

```python
import numpy as np

# Define two vectors x and y
x = np.array([x1, x2, x3])
y = np.array([y1, y2, y3])

# Calculate cosine similarity
cosine_similarity = np.dot(x, y) / (np.linalg.norm(x) * np.linalg.norm(y))

# Calculate cosine distance (1 - cosine similarity)
cosine_distance = 1 - cosine_similarity
print(cosine_distance)


#### Cosine Similarity vs Cosine Distance

Below is a comparison between **Cosine Similarity** and **Cosine Distance** in tabular form:

| **Metric**            | **Cosine Similarity**                                   | **Cosine Distance**                                   |
|-----------------------|---------------------------------------------------------|-------------------------------------------------------|
| **Definition**         | Measures the cosine of the angle between two vectors    | Measures how dissimilar two vectors are               |
| **Formula**            | $(\frac{x \cdot y}{\|x\| \|y\|})$                       | $(1 - \frac{x \cdot y}{\|x\| \|y\|})$                 |
| **Range**              | \([-1, 0, 1]\)                                          | \([0, 1, 2]\)                                         |
| **Interpretation**     | 1: Vectors are identical, 0: Vectors are orthogonal, -1: Vectors are opposite | 0: Vectors are identical, 1: Vectors are orthogonal, 2: Vectors are opposite |
| **Use Case**           | Similarity-based algorithms (e.g., document similarity) | Distance-based algorithms (e.g., clustering)          |

##### Key Notes:
- **Cosine Similarity** focuses on how **similar** two vectors are, based on the angle between them.
- **Cosine Distance** focuses on how **dissimilar** two vectors are, derived by subtracting cosine similarity from 1.


#### Example : distance settings

```python
collection = client.create_collection(
        name="collection_name",
        metadata={"hnsw:space": "cosine"} # l2 is the default
    )
```

`metadata={"hnsw: "cosine"}:`

- **metadata**: This allows you to configure specific settings for the collection.
- **hnsw**: Refers to the search space for the **Hierarchical Navigable Small World (HNSW)** algorithm, which is used for fast approximate nearest neighbor search.
- **"cosine"**: Specifies **cosine similarity** as the distance metric, which measures the angle between two vectors (i.e., how similar they are directionally).
  
By default, ChromaDB uses **L2 (Euclidean distance)**, but here it’s overridden to use **cosine similarity**.

`Cosine Similarity vs L2 Distance:`

- **Cosine Similarity**:
  - Compares the angle between two vectors to measure similarity, ignoring magnitude differences.
  - Suitable for text or document embeddings where direction matters more than vector length.

- **L2 Distance (Euclidean)** (default):
  - Measures the straight-line distance between vectors, and is sensitive to both direction and magnitude.
  - Often used in geometric or spatial data comparisons.


## change the collection settings

- `distance = cosine`

- `Alternatives to "hnsw" in ChromaDB`

In ChromaDB, **"hnsw"** refers to the **Hierarchical Navigable Small World** algorithm, which is commonly used for fast approximate nearest neighbor searches. However, ChromaDB supports other algorithms and methods for vector search and storage.

Here are some alternatives:

1. **"flat"**:
   - The **flat index** performs an exact nearest neighbor search.
   - This is slower compared to "hnsw" but guarantees exact results.
   - Suitable when speed is less of a concern, and accuracy is paramount.

2. **"ivf"** (Inverted File Index):
   - Combines clustering and approximate search.
   - Vectors are grouped into clusters, and searching happens only in the nearest cluster.
   - It’s faster than "flat" for large datasets but provides approximate results.
  
3. **"pq"** (Product Quantization):
   - Compresses vectors into smaller codes, reducing memory usage.
   - Useful for scenarios where memory efficiency is a priority over search accuracy.
   - Offers an approximate nearest neighbor search like "hnsw."

**Summary of Alternatives:**
- **hnsw**: Fast approximate nearest neighbor search (default).
- **flat**: Exact nearest neighbor search with slower performance.
- **ivf**: Clustered approximate search for large datasets.
- **pq**: Compressed vectors for memory efficiency and approximate searches.

**Example Usage:**
```python
metadata={"flat:space": "l2"}   # Uses the flat index with Euclidean distance
metadata={"ivf:space": "cosine"} # Uses IVF with cosine similarity


In [ ]:
collection_cosine = client.get_or_create_collection(name     = "MedicalQA_cosine",
                                                    metadata = {"hnsw:space": "cosine"} # l2 is the default
                                     )

#### Explanation of `collection_cosine.add()`

- **`collection_cosine.add()`**: This function adds data to the ChromaDB collection named `collection_cosine`. The function expects **IDs**, **documents**, and **metadata** for each entry.

Here's a detailed breakdown of each parameter used:

1. **`ids=[str(i) for i in range(0, 100)]`**:
   - Generates a list of 100 IDs as strings from 0 to 99.
   - Each entry in the collection requires a unique ID, which is used to reference or retrieve the document later.
   - The list comprehension `str(i) for i in range(0, 100)` creates these unique string IDs (`'0'`, `'1'`, ..., `'99'`).

2. **`documents=dataset_cleaned["exp"][:100]`**:
   - **`dataset_cleaned["exp"]`** refers to the cleaned dataset containing explanations in the `"exp"` column.
   - The slice `[:100]` selects the first 100 explanation entries.
   - These explanations are stored as **documents** in the collection. The document field usually contains the main content or text you want to create embeddings for.

3. **`metadatas=[{"type": "exp"} for _ in range(0, 100)]`**:
   - A list of metadata for each document.
   - Here, it creates 100 identical metadata dictionaries, each containing `{"type": "exp"}`, signifying that the document type is an explanation (`"exp"`).
   - Metadata provides additional information about the document, which can be useful for categorization, filtering, or retrieval.

#### Summary:
- The `collection_cosine.add()` function stores **100 entries** in the `collection_cosine`.
- Each entry consists of:
  - A unique string **ID** (from 0 to 99),
  - An explanation **document** (from the cleaned dataset),
  - **Metadata** indicating the document type.
  
This prepares the collection to perform vector searches (such as cosine similarity searches) using the added explanations.


In [ ]:
# Suppress warning messages
import logging

from chromadb.config import Settings

# Set up logging configuration at the very beginning

logger = logging.getLogger("chromadb")
logger.setLevel(logging.ERROR)  # Ensure the logger for chromadb is set to ERROR

In [ ]:
client.list_collections()
#client.delete_collection(name="MedicalQA")

In [ ]:
%%time
# Now we will store the expert explanation field of first 10 questions from dataset into collection. 
collection_cosine.add(
    ids       = [str(i) for i in range(0, 100)],                      # Unique string IDs for each entry
    documents = dataset_cleaned["exp"][:100],                         # Explanation field text for the first 100 entries
    metadatas = [{"subject_name": dataset_cleaned["subject_name"][i], 
                  "topic_name":   dataset_cleaned["topic_name"][i]} 
                 for i in range(100)]                                 # Include subject_name and topic_name
)

In [ ]:
for each_query in query_texts:
    results = collection_cosine.query(query_texts=each_query, n_results=1)

    print(each_query)
    print(results['documents'])
    print(results['distances'])

    print()